pddlenv/search/hillclimbing.py

```diff
@@ -0,0 +1,65 @@
+import heapq
+import time
+from typing import AbstractSet, Callable, Dict, Optional, Sequence, Tuple
+
+from pddlenv import env
+from pddlenv.base import Action, Predicate, Problem
+from pddlenv.search import base, utils
+
+Heuristic = Callable[[AbstractSet[Predicate], Problem], float]
+
+
+class HillClimbing:
+
+    def __init__(self, heuristic: Heuristic, logger: Optional[base.Logger] = None, detect_minima: bool = False):
+        self.heuristic = heuristic
+        self.logger = logger
+        self.detect_minima = detect_minima
+
+    def search(self,
+               state: env.EnvState,
+               time_limit: float = None,
+               expansion_limit: int = None) -> Optional[Sequence[Action]]:
+        expanded_states = 0
+        start_time = time.perf_counter()
+
+        # we assume that the dynamics will never change the problem instance
+        problem = state.problem
+        heap = [base.Candidate(0., state)]
+        parents: Dict[env.EnvState, Optional[Tuple[env.EnvState, Action]]] = {state: None}
+        dynamics = env.PDDLDynamics()
+
+        while heap:
+            if time_limit and time_limit <= time.perf_counter() - start_time:
+                break
+            if expansion_limit and expansion_limit <= expanded_states:
+                break
+
+            expanded_states += 1
+            node  = heapq.heappop(heap)
+            state = node.state
+            value = node.heuristic
+            heap = []           # hill climbing --- forget everything
```

Comment:
Why use a heap if it's cleared at every iteration?

Reply:
same as sorting the list
shouldnt be a large overhead because the list should be small

Reply:
Why the need for sorting anything? Couldn't you just keep track of the min while iterating through the successors or am I missing something?

Reply:
For one, I wanted to only minimally modify GBFS so that the difference becomes obvious.
For another, GBFS was already suffiently tested, so less chance of getting a bug.